File: excelium-common/src/main/java/excelium/common/WildcardUtil.java
/*
 * MIT License
 *
 * Copyright (c) 2018 Excelium
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package excelium.common;

/**
 * Collection of wildcard handling utilities
 *
 * @author PhungDucKien
 * @since 2018.04.15
 */
public class WildcardUtil {

    /**
     * Matches wildcard pattern with support for '?' and '*'.
     * <p>
     * '?' Matches any single character.
     * '*' Matches any sequence of characters (including the empty sequence).
     * </p>
     * The matching should cover the entire input string (not partial).
     *
     * @param s the input string
     * @param p the pattern
     * @return the input string matches the pattern or not
     * @see <a href="https://www.quora.com/How-do-I-solve-the-wildcard-matching-problem-in-Java">How do I solve the wildcard matching problem in Java?</a>
     */
    public static boolean isMatch(String s, String p) {
        int i = 0;
        int j = 0;
        int starIndex = -1;
        int iIndex = -1;

        while (i < s.length()) {
            if (j < p.length() && (p.charAt(j) == '?' || p.charAt(j) == s.charAt(i))) {
                ++i;
                ++j;
            } else if (j < p.length() && p.charAt(j) == '*') {
                starIndex = j;
                iIndex = i;
                j++;
            } else if (starIndex != -1) {
                j = starIndex + 1;
                i = iIndex + 1;
                iIndex++;
            } else {
                return false;
            }
        }
        while (j < p.length() && p.charAt(j) == '*') {
            ++j;
        }
        return j == p.length();
    }
}

File: packages/common/src/WildcardUtil.ts
/*
 * Licensed to the Software Freedom Conservancy (SFC) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The SFC licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/**
 * Collection of wildcard handling utilities
 *
 * @author PhungDucKien
 * @since 2018.04.15
 */
export default class WildcardUtil {
  /**
   * Matches wildcard pattern with support for '?' and '*'.
   * <p>
   * '?' Matches any single character.
   * '*' Matches any sequence of characters (including the empty sequence).
   * </p>
   * The matching should cover the entire input string (not partial).
   *
   * @param s the input string
   * @param p the pattern
   * @return the input string matches the pattern or not
   * @see <a href="https://www.quora.com/How-do-I-solve-the-wildcard-matching-problem-in-Java">How do I solve the wildcard matching problem in Java?</a>
   */
  public static isMatch(s: string, p: string): boolean {
    let i = 0;
    let j = 0;
    let starIndex = -1;
    let iIndex = -1;

    while (i < s.length) {
      if (j < p.length && (p.charAt(j) === '?' || p.charAt(j) === s.charAt(i))) {
        ++i;
        ++j;
      } else if (j < p.length && p.charAt(j) === '*') {
        starIndex = j;
        iIndex = i;
        j++;
      } else if (starIndex !== -1) {
        j = starIndex + 1;
        i = iIndex + 1;
        iIndex++;
      } else {
        return false;
      }
    }
    while (j < p.length && p.charAt(j) === '*') {
      ++j;
    }
    return j === p.length;
  }
}
